includes/islandora_book_batch.inc

```diff
@@ -466,15 +466,30 @@ class IslandoraBookPageBatchObject extends IslandoraFlatBatchObject {
    */
   public function addRelationships() {
     module_load_include('inc', 'islandora_paged_content', 'includes/utilities');
-    // Add relationship to collection.
+    // Add relationship to parent.
     $rels_ext = $this->relationships;
     islandora_paged_content_set_relationship($rels_ext, ISLANDORA_RELS_EXT_URI, 'isPageOf', $this->parentId);
     islandora_paged_content_set_relationship($rels_ext, ISLANDORA_RELS_EXT_URI, 'isSequenceNumber', (string) $this->sequenceNumber, TRUE);
     islandora_paged_content_set_relationship($rels_ext, ISLANDORA_RELS_EXT_URI, 'isPageNumber', (string) $this->sequenceNumber, TRUE);
     islandora_paged_content_set_relationship($rels_ext, ISLANDORA_RELS_EXT_URI, 'isSection', '1', TRUE);
     islandora_paged_content_set_relationship($rels_ext, FEDORA_RELS_EXT_URI, 'isMemberOf', $this->parentId);
+
     // Add content model relationship.
-    $this->models = 'islandora:pageCModel';
+    $child_models = array();
+    $paged_cmodels = islandora_paged_content_retrieve_applicable_cmodels();
+    foreach ($this->preprocessorParameters['content_models'] as $parent_model) {
```

Comment:
So let me see if i understand this: for each given parent CMODEL passed, check which child cmodels are allowed and apply all of them, and also check everytime? What is the benefit here of having many cmodels for a child instead of a single one given a certain parent, which in this case i guess its always the same? Also, would it not be faster to do the following: on "batch submit" calculate this once and pass around?

Reply:
I'm just taking into account we allow multiple models here: https://github.com/Islandora/islandora_book_batch/blob/7.x/islandora_book_batch.drush.inc#L139 and https://github.com/Islandora/islandora_batch/blob/7.x/islandora_batch.drush.inc#L143 seems to be a trend.

I can move the loop to here https://github.com/Islandora/islandora_book_batch/blob/7.x/includes/islandora_book_batch.inc#L434 and only run it once if you agree

Reply:
I think you could pass the child cmodels as ->preprocessor parameters, kinda here https://github.com/Islandora/islandora_book_batch/blob/7.x/includes/islandora_book_batch.inc#L392 (or some place similar but not per child, but calculated on the parent ). Means you could, for each parent CMODEL calculate the children ones and build a hash/map array with each parent CMODEL as key (you have really just one here right?) with a value of children cmodels in a list (array). That way the logic on the child itself  (what you have right now) becomes basically a single line, not invoking other functions nor iterating, just assigning from that preprocessor . Does that make sense?